### DISCUSSION TITLE
Objective Quality Evidence for Unit Test Derivations

### DISCUSSION THREAD

**Discussion by Chad-Peterson** [UIUC-ESDL/SPI2py#6] [General]
Hi Prof @cmccomb, 

Once break is over, I'm going to make a concerted effort with undergrads to write unit tests. For calculations (e.g., rotating a 3D object about a point) I like to draw examples by hand, label points, etc. I think it would be very helpful to capture this history in the repo.

For instance, if I am confused about a test someone else wrote, instead of trying to interpret an "assert x = expected_value" I can see exactly where the "expected_value" came from, and if I agree with the logic or not.

However, I think keeping one PPT slide deck in the test folder (which I'm currently doing) won't work well when many people are editing it. Do you have any thoughts for how to capture this design history? Would PNG/JPG files in each test subfolder work best? I'd prefer to avoid linking to an external google doc or anything. @

**Comment by cmccomb** [comment]
Hey @Chad-Peterson! Please feel free to drop the Prof title. I think that there are a few different approaches to addressing this question:
1. No additional infrastructure - Use descriptive variable names and comment extensively
2. Some additional infrastructure - Integrate visualizations with the tests, so that users can visually audit tests if they fail. 
3. Lots of additional infrastructure - Use a combination of [sphinx](https://packaging.python.org/en/latest/tutorials/creating-documentation/) and [doctest](https://docs.python.org/3/library/doctest.html) to (1) automatically generate documentation based on carefully formatted comments, (2) expose tests as example, and (3) automatically generate images/visualizations that can be integrated into the documentation.

My recommendation would be to take these on in order - focus on robust documentation in your current tests, then add some viz, then integrate it into compiled documentation that can be exposed to users. What do you think?

**Reply by Chad-Peterson** [reply]
Sounds good--one thing I haven't quite gotten a hand on coming back to academia is who prefers what title. I like that approach and forgot we could use something like Matplotlib for the documentation too (I can see how Scipy does it). 

I think for now I will create a shared ppt outside of GitHub (so multiple people can edit it live rather than trying to commit changes one at a time). 

For my unit tests I'm sketching or calculating things by hand to come up with expected values. So I'll upload those sketches to the PPT to save the history without any additional work. Then when SPI2Py becomes more mature we can assign someone with the issue of creating the visualization code.

**Comment by Chad-Peterson** [comment]
Now that I think about it, you can attach documents when replying to issues. I can just specify in the issue for people to attach their work.
